### PR TITLE
add SG-MCMC methods

### PR DESCRIPTION
**Pull request by master**
Add SGHMC and Cyclical SGLD methods along with step size schedulers, preconditioners, and convergence diagnostics toolings.

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

### CODE REVIEW

File: fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_approximator.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from fortuna.prob_model.posterior.sgmcmc.cyclical_sgld import (
+    CYCLICAL_SGLD_NAME,
+)
+
+
+class CyclicalSGLDPosteriorApproximator(PosteriorApproximator):
+    def __init__(
+        self,
+        n_samples: int = 10,
+        n_thinning: Optional[int] = 100,
```
[the remark this answers is not in the document]
Comment (gianlucadetommaso) [reply fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_approximator.py:17]: Also, is 100 a reasonable default? Or would it be too high for expensive models? This is just a question; feel free to convince me that 100 is good.

File: fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_callback.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        return ((self.current_step % self.cycle_length) / self.cycle_length) < self.exploration_ratio \
+            and (self.current_step % self.cycle_length) % self.n_thinning == 0 \
+            and self.samples_count < self.n_samples
+
+    def training_step_end(self, state: TrainState) -> TrainState:
+        self.current_step += 1
+
+        if self.do_sample():
+            if self.save_checkpoint_dir:
+                self.trainer.save_checkpoint(
```
Comment (gianlucadetommaso) [inline fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_callback.py:81]: Do you need `save_checkpoint`? It seem you are anyway putting the state in the state repository. If a `checkpoint_dir` is passed o the repository, the state is anyway saved on disk. 

Am I missing something?

File: fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_integrator.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+        updates, state = jax.lax.cond(
+            ((state.sgld_state.count % cycle_length) / cycle_length)
+            >= exploration_ratio,
+            sgld_step,
+            sgd_step,
+        )
+        return updates, state
+
+    return GradientTransformation(init_fn, update_fn)
```
Comment (gianlucadetommaso) [inline fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_integrator.py:104]: Very neat!

**Review by gianlucadetommaso — COMMENTED** [review]
An initial set of comment. I will be going forward on the PR soon.

I can already say: great work!

File: fortuna/prob_model/posterior/sgmcmc/sghmc/sghmc_posterior.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+    def __str__(self):
+        return SGHMC_NAME
+
+    def fit(
+        self,
+        train_data_loader: DataLoader,
+        val_data_loader: Optional[DataLoader] = None,
+        fit_config: FitConfig = FitConfig(),
+        map_fit_config: Optional[FitConfig] = None,
```
Comment (gianlucadetommaso) [inline fortuna/prob_model/posterior/sgmcmc/sghmc/sghmc_posterior.py:64]: Here you do use the preliminary run with MAP. How come here it appears but not in Cyclical SGLD? I guess I missed something before.

File: fortuna/prob_model/posterior/sgmcmc/sghmc/sghmc_posterior.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            n_epochs=fit_config.optimizer.n_epochs,
+            metrics=fit_config.monitor.metrics,
+            validation_dataloader=val_data_loader,
+            validation_dataset_size=val_data_loader.size if val_data_loader is not None else None,
+            verbose=fit_config.monitor.verbose,
+            callbacks=[sghmc_sampling_callback],
+        )
+        logging.info("Fit completed.")
+
+        if sghmc_sampling_callback.samples_count < self.posterior_approximator.n_samples:
```
Comment (gianlucadetommaso) [inline fortuna/prob_model/posterior/sgmcmc/sghmc/sghmc_posterior.py:163]: Same comment as above.

File: fortuna/prob_model/posterior/sgmcmc/sgmcmc_diagnostic.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    samples = ravel_pytree(samples)[0].reshape(len(samples), -1)
+    grads = ravel_pytree(grads)[0].reshape(len(grads), -1)
+
+    def _k_0(param1, param2, grad1, grad2, c, beta):
+        dim = param1.shape[0]
+        diff = param1 - param2
+        base = c**2 + jnp.dot(diff, diff)
+        kern = jnp.dot(grad1, grad2) * base**beta
+        kern += -2 * beta * jnp.dot(grad1, diff) * base ** (beta - 1)
+        kern += 2 * beta * jnp.dot(grad2, diff) * base ** (beta - 1)
```
Comment (gianlucadetommaso) [inline fortuna/prob_model/posterior/sgmcmc/sgmcmc_diagnostic.py:45]: It looks like the `base` always goes at the denominator, since `beta` is negative. For example, with the default `beta=-0.5`, we have `... / base ** (3/2)`.

Just want to check, in case this is not the intended behaviour.

File: fortuna/prob_model/posterior/sgmcmc/sgmcmc_diagnostic.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        return kern
+
+    _batched_k_0 = vmap(_k_0, in_axes=(None, 0, None, 0, None, None))
+
+    def _ksd(accum, x):
+        sample1, grad1 = x
+        accum += jnp.sum(_batched_k_0(sample1, samples, grad1, grads, c, beta))
+        return accum, None
+
+    ksd_sum, _ = lax.scan(_ksd, 0.0, (samples, grads))
```
Comment (gianlucadetommaso) [inline fortuna/prob_model/posterior/sgmcmc/sgmcmc_diagnostic.py:63]: If we take one sample at the time, likely you might have trouble using `scan` here. To make it work, consider using the `pure_callback`, like in `DeepEnsemblePosterior.sample`.

File: fortuna/prob_model/posterior/sgmcmc/sgmcmc_posterior_mixin.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            A sample from the posterior distribution.
+        """
+        if rng is None:
+            rng = self.rng.get()
+        state = pure_callback(
+            lambda j: self.state.get(i=j),
+            self.state.get(i=0),
+            random.choice(rng, self.posterior_approximator.n_samples),
+        )
+        return JointState(
```
Comment (gianlucadetommaso) [inline fortuna/prob_model/posterior/sgmcmc/sgmcmc_posterior_mixin.py:47]: Why `JointState` and not `SGMCMCState`? Just trying to understand.

**Review by gianlucadetommaso — CHANGES_REQUESTED** [review]
Fantastic work! Here a few comments and questions. Let me know what you think!

File: fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_callback.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        trainer: TrainerABC
+            An instance of the trainer class.
+        state_repository: TrainStateRepository
+            An instance of the state repository.
+        keep_top_n_checkpoints: int
+            Number of past checkpoint files to keep.
+        save_checkpoint_dir: Optional[Path]
+            The optional path to save checkpoints.
+        """
+        if n_samples * cycle_length < cycle_length != 0:
```
[the remark this answers is not in the document]
Comment (master) [reply fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_callback.py:54]: Yes :)

File: docs/source/references/prob_model/posterior/sgmcmc.rst [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+Preconditioners
+===============
+
+Fortuna provides implementations of preconditioners to improve samplers efficacy.
+
+.. automodule:: fortuna.prob_model.posterior.sgmcmc.sgmcmc_preconditioner
+    :exclude-members: Preconditioner, PreconditionerState, RMSPropPreconditionerState, IdentityPreconditionerState
+
+
+Diagnostics
```
[the remark this answers is not in the document]
Comment (master) [reply docs/source/references/prob_model/posterior/sgmcmc.rst:73]: Yes, that's something to think about. Other libraries (e.g., `blackjax`) typically have MCMC and SG-MCMC in separate modules, and diagnostics on a high level. Perhaps we can adopt the same structure? Or, alternatively, treat SG-MCMC as a special case of MCMC, and have something like `posterior.mcmc.hmc`, `posterior.mcmc.sghmc`, `posterior.mcmc.cyclical_sgld`, etc?

File: fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_posterior.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+            n_epochs=fit_config.optimizer.n_epochs,
+            metrics=fit_config.monitor.metrics,
+            validation_dataloader=val_data_loader,
+            validation_dataset_size=val_data_loader.size if val_data_loader is not None else None,
+            verbose=fit_config.monitor.verbose,
+            callbacks=[cyclical_sampling_callback],
+        )
+        logging.info("Fit completed.")
+
+        if cyclical_sampling_callback.samples_count < self.posterior_approximator.n_samples:
```
[the remark this answers is not in the document]
Comment (master) [reply fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_posterior.py:147]: Yes, I think I'll just refactor it to compute the resulting number of samples given the size of `train_data` and `n_thinning` / `cycle_length` parameters.

File: fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_posterior.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+            eval_every_n_epochs=fit_config.monitor.eval_every_n_epochs,
+            early_stopping_monitor=fit_config.monitor.early_stopping_monitor,
+            early_stopping_min_delta=fit_config.monitor.early_stopping_min_delta,
+            early_stopping_patience=fit_config.monitor.early_stopping_patience,
+        )
+
+        if super()._is_state_available_somewhere(fit_config):
+            state = self._restore_state_from_somewhere(fit_config=fit_config)
+        else:
+            state = self._init_state(
```
[the remark this answers is not in the document]
Comment (master) [reply fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_posterior.py:105]: Right, I was going to omit preliminary MAP from Cyclical SGLD because it performs SGD steps in the beginning of each cycle, but forgot about the partial freezing / inference on a subset of parameters use case. Will add it back.

File: docs/source/references/prob_model/posterior/sgmcmc.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (gianlucadetommaso) [reply docs/source/references/prob_model/posterior/sgmcmc.rst:73]: Which option do you think is best? I think the same structure will be applied to the code. I don't have a strong preference here.

File: fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_approximator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (master) [reply fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_approximator.py:17]: My idea was to support both `None` and 1 as a "disable" value: https://github.com/master/fortuna/blob/sgmcmc/fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_approximator.py#L42

Regarding 100, I don't think there's a one-size-fits-all default, because it really depends on the training budget. OTOH, we already have the default deep ensemble size, which are also model-dependent, so it'd be good to have some sensible value here as well. Perhaps just setting `n_thinning = 1` for now?

[the remark this answers is not in the document]
Comment (gianlucadetommaso) [reply fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_approximator.py:17]: Makes sense. If in your experiments you find better defaults, we can change it back!

File: fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_posterior.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        trainer_cls = select_trainer_given_devices(
+            devices=fit_config.processor.devices,
+            BaseTrainer=MAPTrainer,
+            JittedTrainer=JittedMAPTrainer,
+            MultiDeviceTrainer=MultiDeviceMAPTrainer,
+            disable_jit=fit_config.processor.disable_jit,
+        )
+
+        save_checkpoint_dir = (
+            pathlib.Path(fit_config.checkpointer.save_checkpoint_dir) / "c"
```
[the remark this answers is not in the document]
Comment (master) [reply fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_posterior.py:86]: "c" is for "chain". I'm currently checkpointing the chain alongside with samples in the state repository. However, that's likely redundant especially for the cyclical method.

File: fortuna/prob_model/posterior/sgmcmc/sgmcmc_diagnostic.py (hunk 6 above)
Comment (master) [reply fortuna/prob_model/posterior/sgmcmc/sgmcmc_diagnostic.py:45]: It's a re-implementation of the reference code and I hope it's correct: https://github.com/jgorham/SteinDiscrepancy.jl/blob/master/src/kernels/SteinInverseMultiquadricKernel.jl (notice that they [flipped the sign in the parametrization](https://github.com/jgorham/SteinDiscrepancy.jl/blob/master/src/kernels/SteinInverseMultiquadricKernel.jl#L8)).

File: docs/source/references/prob_model/posterior/sgmcmc.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (master) [reply docs/source/references/prob_model/posterior/sgmcmc.rst:73]: I think I'd opt for having `mcmc` as a top-level package, with full-batch and stochastic methods, along with diagnostics under the same umbrella. After all, there is nothing minibatch-specific in KSD - it can be applied to samples generated with any approach. Just need to make it clear in the docs to not confuse batched HMC with stochastic gradients with SGHMC that is SGLD with momentum.

File: fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_callback.py (hunk 2 above)
Comment (master) [reply fortuna/prob_model/posterior/sgmcmc/cyclical_sgld/cyclical_sgld_callback.py:81]: This is similar to https://github.com/awslabs/fortuna/blob/main/fortuna/prob_model/posterior/deep_ensemble/deep_ensemble_posterior.py#L106-L110 in DeepEnsembles: the state is a collection of checkpoints, which are getting saving even if `dump_state` is False. We can throw a warning about that to inform the user.

File: fortuna/prob_model/posterior/sgmcmc/sgmcmc_posterior_mixin.py (hunk 8 above)
Comment (master) [reply fortuna/prob_model/posterior/sgmcmc/sgmcmc_posterior_mixin.py:47]: It's a mixin for `Posterior` subclasses, so following the contract here: https://github.com/awslabs/fortuna/blob/main/fortuna/prob_model/posterior/base.py#L125 Maybe it would be clearer to refactor it into a base class instead of mixin?

**Review by gianlucadetommaso — COMMENTED** [review]
So the parallel build was the problem eventually?
